File: Difficulty: Medium/Fraction Trouble/fraction-trouble.js
//{ Driver Code Starts
//Initial Template for javascript

//Initial Template for javascript
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => {
        return string.trim();
    });
    
    main();    
});

function readLine() {
    return inputString[currentLine++];
}

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for(;i<t;i++)
    {
        let input_ar0 = readLine().split(' ').map(x=>parseInt(x));
        let n = input_ar0[0];
        let d = input_ar0[1];
        
        let obj = new Solution();
        let arr = obj.LargestFraction(n, d);
        let ans = "";
        for(let element of arr) ans += element + " ";
        
        console.log(ans);
    
console.log("~");
}
}
// } Driver Code Ends

//User function Template for javascript

class Solution {
     // Function to calculate the GCD of two numbers
     gcd(a, b) {
        a = Math.abs(a);
        b = Math.abs(b);
        
        // Ensure a is the larger number for the GCD calculation
        if (b > a) {
            var temp = a;
            a = b;
            b = temp;
        }
        
        // Use the Euclidean algorithm to find the GCD
        while (true) {
            if (b === 0) return a;  // If b is 0, a is the GCD
            a %= b;  // Reduce a by remainder of a divided by b
            if (a === 0) return b;  // If a is 0, b is the GCD
            b %= a;  // Reduce b by remainder of b divided by a
        }
    }
    
    // Function to find the largest fraction with numerator n and denominator d
    LargestFraction(n, d) {
        let r = 0, s = 1;  // Initialize result as a fraction 0/1
        let limit = 10000;  // Set a limit for the denominator q
        
        // Iterate through possible values of q from the limit down to 2
        for (let q = limit; q >= 2; q--) {
            // Calculate the numerator p for the fraction p/q
            let p = Math.floor((n * q - 1) / d);
            
            // Check if the fraction p/q is larger than the current best fraction r/s
            if (p * s >= r * q) {
                s = q;  // Update the denominator
                r = p;  // Update the numerator
            }
        }
        
        // Find the GCD of r and s to simplify the fraction
        let D = this.gcd(r, s);
        let res = [];
        
        // Add the simplified numerator and denominator to the result array
        res.push(Math.floor(r / D));
        res.push(Math.floor(s / D));
        
        return res;  // Return the simplified fraction
    }
}

File: Difficulty: Easy/Fraction Trouble/fraction-trouble.cpp
//{ Driver Code Starts
// Initial Template for C++
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends

// User function Template for C++
class Solution {
  public:
    typedef long double ld;
    
    // Function to find a pair of numbers (i, j) where i > n and j > d, such that
    // i/j is the closest but smaller than n/d
    vector<int> numAndDen(int n, int d) {
        // Calculate the value of n/d
        double val = (double)n / d;
        
        // Initialize max to store the closest value smaller than n/d
        double max = -1;
        
        // Initialize the result vector to store the final pair
        vector<int> res;
        
        // Loop through values of i greater than n
        for (int i = n + 1; i <= 10000; i++) {
            // Calculate the corresponding value of j using the formula
            int j = (i * d) / n + 1;
            
            // If j exceeds 10000, break the loop
            if (j > 10000) {
                break;
            }
            
            // Calculate the value of i/j
            double v = (double)i / j;
            
            // Check if this value is smaller than n/d but larger than the previous max
            if (v < val && v > max) {
                max = v;
                // Update the result pair with i and j
                res = {i, j};
            }
        }
        
        // Return the result vector containing the pair (i, j)
        return res;
    }
};



//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        int n,d;
        cin >> n >> d;
        Solution ob;
        vector<int>ans=ob.numAndDen(n,d);
        for(int i=0;i<ans.size();i++)cout<<ans[i]<<" ";
        cout<<endl;
    
cout << "~" << "\n";
}
    return 0;
}

// } Driver Code Ends